File: brain/verse.py
"""Respect verse start/stop when picking a mix-in cue or a mix-out ride.

Canonical logic lives in `core-rust/clawdj/src/verse.rs` (`clawdj verse cue`).
This module is the in-process planner copy so mix builds do not spawn cargo.
Keep the tests in lockstep with the Rust unit tests.
"""
from __future__ import annotations

import json
from dataclasses import asdict, dataclass
from typing import Any, Iterable

START_TOLERANCE_SECONDS = 2.0
INTRO_BEFORE_VERSE_SECONDS = 8.0
DEFAULT_BLEND_BEATS = 32
# Detector sometimes labels a whole cut as one "verse". Real rap/sung
# verses are ~16–24 bars. Do not auto-extend into that garbage.
MAX_VERSE_SECONDS = 80.0


@dataclass(frozen=True)
class LyricSegment:
    kind: str
    start: float
    end: float


@dataclass
class VerseCueDecision:
    legal: bool
    placement: str
    cue_seconds: float
    reason: str
    source: str

    def as_dict(self) -> dict[str, Any]:
        return asdict(self)


def _segments(raw: Iterable[dict[str, Any] | LyricSegment]) -> list[LyricSegment]:
    out: list[LyricSegment] = []
    for item in raw or []:
        if isinstance(item, LyricSegment):
            out.append(item)
            continue
        kind = str(item.get("kind") or "")
        try:
            start = float(item.get("start"))
            end = float(item.get("end"))
        except (TypeError, ValueError):
            continue
        if kind:
            out.append(LyricSegment(kind=kind, start=start, end=end))
    return out


def _is_verse(kind: str) -> bool:
    return kind.casefold() == "verse"


def _is_chorus(kind: str) -> bool:
    return kind.casefold() in {"chorus", "hook"}


def _containing(cue: float, segments: list[LyricSegment]) -> LyricSegment | None:
    # A boundary timestamp belongs to the segment that *starts* there
    # (chorus end 27.75 / verse start 27.75 is the verse).
    for segment in segments:
        if abs(segment.start - cue) <= 1e-9:
            return segment
    for segment in segments:
        if cue + 1e-9 > segment.start and cue < segment.end - 1e-9:
            return segment
    for segment in segments:
        if cue + 1e-9 >= segment.start and cue <= segment.end + 1e-9:
            return segment
    return None


def classify_cue(cue: float, segments: Iterable[dict[str, Any] | LyricSegment]) -> str:
    """Where a cue sits: intro / verse_start / mid_verse / chorus / mid_chorus / unknown."""
    items = _segments(segments)
    if not items:
        return "unknown"
    if cue <= START_TOLERANCE_SECONDS:
        first = min(items, key=lambda item: item.start)
        if first.start >= INTRO_BEFORE_VERSE_SECONDS or cue + START_TOLERANCE_SECONDS < first.start:
            return "intro"
    hit = _containing(cue, items)
    if hit is None:
        first_start = min(item.start for item in items)
        if cue + START_TOLERANCE_SECONDS < first_start:
            return "intro"
        return "unknown"
    into = cue - hit.start
    if _is_verse(hit.kind):
        return "verse_start" if into <= START_TOLERANCE_SECONDS else "mid_verse"
    if _is_chorus(hit.kind):
        return "chorus" if into <= START_TOLERANCE_SECONDS else "mid_chorus"
    if cue <= START_TOLERANCE_SECONDS:
        return "intro"
    return "unknown"


def has_vocal_verses(title: str = "", path: str = "") -> bool:
    """False for instrumental-only / no-vocal beds — no artist verse exists."""
    from brain.stems import classify_stem

    return classify_stem(title or "", path or "") != "instrumental_only"


def _has_intro_before(verse: LyricSegment, items: list[LyricSegment]) -> bool:
    return verse.start >= INTRO_BEFORE_VERSE_SECONDS or any(
        item.end <= verse.start + 1e-9 and item.start < verse.start for item in items
    )


def verse_cut_by_window(
    start: float,
    end: float,
    segments: Iterable[dict[str, Any] | LyricSegment],
) -> LyricSegment | None:
    """First real verse whose interior is eaten by [start, end]."""
    items = _segments(segments)
    for item in items:
        if not _is_verse(item.kind):
            continue
        if (item.end - item.start) > MAX_VERSE_SECONDS:
            continue
        overlap_lo = max(start, item.start + START_TOLERANCE_SECONDS)
        overlap_hi = min(end, item.end)
        if overlap_hi > overlap_lo + 1e-9:
            return item
    return None


def respect_verse_entry(
    proposed_cue: float,
    segments: Iterable[dict[str, Any] | LyricSegment],
    *,
    first_beat: float = 0.0,
    bpm: float | None = None,
    blend_beats: int = DEFAULT_BLEND_BEATS,
    title: str = "",
    path: str = "",
) -> VerseCueDecision:
    """Rewrite a cue that is mid-verse, or whose incoming blend *finishes* mid-verse."""
    del first_beat  # reserved: Mixxx first-beat offset is not a mid-verse landing
    if title or path:
        if not has_vocal_verses(title, path):
            return VerseCueDecision(
                legal=True,
                placement="unknown",
                cue_seconds=max(0.0, float(proposed_cue)),
                reason="instrumental / no vocals — verse boundaries do not apply",
                source="unchanged",
            )
    items = _segments(segments)
    placement = classify_cue(proposed_cue, items)
    period = 60.0 / bpm if bpm and bpm > 0 else None
    fader_at = (
        float(proposed_cue) + max(1, int(blend_beats)) * period if period else float(proposed_cue)
    )
    if float(proposed_cue) <= START_TOLERANCE_SECONDS:
        return VerseCueDecision(
            legal=True,
            placement=placement if placement != "unknown" else "intro",
            cue_seconds=max(0.0, float(proposed_cue)),
            reason="starting from the top is a legal mix-in",
            source="unchanged",
        )
    cut = verse_cut_by_window(float(proposed_cue), fader_at, items)
    if cut is None and placement != "mid_verse":
        return VerseCueDecision(
            legal=True,
            placement=placement,
            cue_seconds=max(0.0, float(proposed_cue)),
            reason=f"{placement} is a legal mix-in",
            source="unchanged",
        )
    verse = cut or _containing(proposed_cue, items)
    if verse is None or not _is_verse(verse.kind):
        return VerseCueDecision(
            legal=True,
            placement=placement,
            cue_seconds=max(0.0, float(proposed_cue)),
            reason="mid-verse but no containing verse segment",
            source="unchanged",
        )
    if _has_intro_before(verse, items):
        return VerseCueDecision(
            legal=False,
            placement="mid_verse",
            cue_seconds=0.0,
            reason=(
                f"incoming blend from {proposed_cue:.2f}s finishes inside verse "
                f"{verse.start:.2f}–{verse.end:.2f}; start from 0:00 so the blend "
                "covers the intro and the verse starts after the fader"
            ),
            source="intro_top",
        )
    preroll = (max(1, int(blend_beats)) * period) if period else 0.0
    cue = max(0.0, verse.start - preroll)
    return VerseCueDecision(
        legal=False,
        placement="mid_verse",
        cue_seconds=cue,
        reason=(
            f"incoming blend from {proposed_cue:.2f}s finishes inside verse "
            f"{verse.start:.2f}–{verse.end:.2f}; pre-roll onto verse start"
        ),
        source="verse_preroll",
    )


def respect_verse_exit(
    cue_seconds: float,
    ride_beats: int,
    bpm: float,
    segments: Iterable[dict[str, Any] | LyricSegment],
    *,
    blend_beats: int = DEFAULT_BLEND_BEATS,
    title: str = "",
    path: str = "",
) -> tuple[int, str]:
    """Extend a ride so the whole outgoing blend sits after the current verse."""
    if title or path:
        if not has_vocal_verses(title, path):
            return int(ride_beats), "unchanged"
    items = _segments(segments)
    if not items or bpm <= 0:
        return int(ride_beats), "unchanged"
    period = 60.0 / bpm
    fade_at = float(cue_seconds) + int(ride_beats) * period
    fade_done = fade_at + max(1, int(blend_beats)) * period
    verse = verse_cut_by_window(fade_at, fade_done, items)
    if verse is None:
        return int(ride_beats), "unchanged"
    needed = max(0, int((verse.end - float(cue_seconds)) / period + 0.999))
    if needed <= int(ride_beats):
        return int(ride_beats), "unchanged"
    capped = min(needed, int(ride_beats) + 128)
    return capped, (
        f"blend {fade_at:.2f}–{fade_done:.2f}s eats verse {verse.start:.2f}–{verse.end:.2f}; "
        f"extend ride {int(ride_beats)} -> {capped} so the fade starts at the verse end"
    )


@dataclass
class VerseViolation:
    track: str
    side: str
    at_seconds: float
    placement: str
    reason: str


def audit_mix_plan(
    plan: dict[str, Any],
    segments_by_id: dict[str, list[dict[str, Any]]],
) -> list[VerseViolation]:
    """Report mix-to-listen verse cuts. Deterministic; no subagent."""
    by_id = {str(row.get("track_id")): row for row in plan.get("tracks") or []}
    bodies = {
        str(event.get("track") or ""): event
        for event in plan.get("events") or []
        if event.get("op") == "play_body"
    }
    transitions = [event for event in plan.get("events") or [] if event.get("op") == "transition"]
    violations: list[VerseViolation] = []
    for event in transitions:
        to_label = str(event.get("to_track") or "")
        from_label = str(event.get("from_track") or "")
        blend = int(event.get("transition_beats") or DEFAULT_BLEND_BEATS)
        incoming = next(
            (
                row
                for row in by_id.values()
                if f"{row.get('artist')} — {row.get('title')}" == to_label
            ),
            None,
        )
        outgoing = next(
            (
                row
                for row in by_id.values()
                if f"{row.get('artist')} — {row.get('title')}" == from_label
            ),
            None,
        )
        if incoming and event.get("technique") != "vocal_over_bed":
            if not has_vocal_verses(
                str(incoming.get("title") or ""),
                str(incoming.get("track_id") or ""),
            ):
                incoming = None
        if incoming and event.get("technique") != "vocal_over_bed":
            segs = segments_by_id.get(str(incoming.get("track_id"))) or []
            cue = float(incoming.get("cue_seconds") or 0.0)
            bpm = float(incoming.get("bpm") or 0.0)
            period = 60.0 / bpm if bpm > 0 else 0.0
            fader_at = cue + blend * period
            # Cue 0 is a legal iconic intro even if the detector labeled the
            # first shouts as "verse".
            cut = (
                None
                if cue <= START_TOLERANCE_SECONDS
                else verse_cut_by_window(cue, fader_at, segs)
            )
            if cut is not None:
                violations.append(
                    VerseViolation(
                        track=to_label,
                        side="entry",
                        at_seconds=round(fader_at, 2),
                        placement="mid_verse",
                        reason=(
                            f"cue {cue:.2f}s + {blend}-beat blend finishes at "
                            f"{fader_at:.2f}s inside verse {cut.start:.2f}–{cut.end:.2f}"
                        ),
                    )
                )
        if outgoing and not has_vocal_verses(
            str(outgoing.get("title") or ""),
            str(outgoing.get("track_id") or ""),
        ):
            outgoing = None
        if outgoing:
            segs = segments_by_id.get(str(outgoing.get("track_id"))) or []
            body = bodies.get(from_label) or {}
            if body.get("trust_ride_beats"):
                continue
            cue = float(outgoing.get("cue_seconds") or 0.0)
            bpm = float(outgoing.get("bpm") or 0.0)
            ride = int(body.get("beats") or 0)
            if bpm <= 0 or ride <= 0:
                continue
            period = 60.0 / bpm
            fade_at = cue + ride * period
            fade_done = fade_at + blend * period
            cut = verse_cut_by_window(fade_at, fade_done, segs)
            if cut is not None:
                violations.append(
                    VerseViolation(
                        track=from_label,
                        side="exit",
                        at_seconds=round(fade_at, 2),
                        placement="mid_verse",
                        reason=(
                            f"blend {fade_at:.2f}–{fade_done:.2f}s eats verse "
                            f"{cut.start:.2f}–{cut.end:.2f}"
                        ),
                    )
                )
    return violations


def segments_from_lyrics_json(artist: str, title: str) -> list[dict[str, Any]]:
    """Verse/chorus map from the on-disk LRCLIB cache when sqlite is empty."""
    from brain.lyric_timeline import detect_segments, parse_lrc
    from brain.lyrics import LYRICS_DIR, cache_key

    path = LYRICS_DIR / f"{cache_key(artist, title)}.json"
    if not path.exists():
        return []
    try:
        data = json.loads(path.read_text())
    except (OSError, json.JSONDecodeError):
        return []
    lrc = data.get("synced_lyrics") or data.get("syncedLyrics") or data.get("lrc") or ""
    if not lrc:
        return []
    return [
        {"kind": item.kind, "start": item.start, "end": item.end, "first_line": item.first_line}
        for item in detect_segments(parse_lrc(lrc))
    ]


def fill_segment_lookup(
    tracks: list[dict[str, Any]], lookup: dict[str, list[dict[str, Any]]]
) -> dict[str, list[dict[str, Any]]]:
    """Sqlite segments win; lyrics JSON fills gaps so verse_guard actually runs."""
    filled = dict(lookup)
    for row in tracks:
        track_id = str(row.get("track_id") or "")
        if not track_id:
            continue
        if not has_vocal_verses(str(row.get("title") or ""), track_id):
            filled.pop(track_id, None)
            continue
        if filled.get(track_id):
            continue
        extra = segments_from_lyrics_json(str(row.get("artist") or ""), str(row.get("title") or ""))
        if extra:
            filled[track_id] = extra
    return filled

File: core-rust/clawdj/src/verse.rs
//! Respect verse start/stop when picking a mix-in cue or a mix-out ride.
//!
//! Automatic energy cues (`phrase_body`) land ~30–45s in — usually mid
//! first verse. A legal entry is the top of the record or a pre-roll that
//! finishes on bar 1 of the verse. A legal exit starts the blend at the
//! verse's end, not in the middle of a rap or sung verse.

use serde::{Deserialize, Serialize};

use crate::stems::{StemKind, classify_stem};

/// How far past a segment start still counts as "on the 1".
pub const START_TOLERANCE_SECONDS: f64 = 2.0;
/// A verse that begins after this has intro/hook before it.
pub const INTRO_BEFORE_VERSE_SECONDS: f64 = 8.0;
/// Detector sometimes labels a whole cut as one verse. Skip those.
pub const MAX_VERSE_SECONDS: f64 = 80.0;

/// One lyric-timeline segment (verse or chorus).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LyricSegment {
    pub kind: String,
    pub start: f64,
    pub end: f64,
}

/// Where a cue sits relative to lyric structure.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CuePlacement {
    Intro,
    VerseStart,
    MidVerse,
    Chorus,
    MidChorus,
    Unknown,
}

/// Rewrite of an illegal automatic cue.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VerseCueDecision {
    pub legal: bool,
    pub placement: CuePlacement,
    pub cue_seconds: f64,
    pub reason: String,
    /// `unchanged` | `intro_top` | `verse_preroll`
    pub source: String,
}

/// JSON body for `clawdj verse cue`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VerseCueRequest {
    pub proposed_cue: f64,
    #[serde(default)]
    pub bpm: Option<f64>,
    #[serde(default)]
    pub first_beat: Option<f64>,
    #[serde(default = "default_blend_beats")]
    pub blend_beats: u32,
    #[serde(default)]
    pub segments: Vec<LyricSegment>,
    #[serde(default)]
    pub title: String,
    #[serde(default)]
    pub path: String,
}

fn default_blend_beats() -> u32 {
    32
}

fn is_verse(kind: &str) -> bool {
    kind.eq_ignore_ascii_case("verse")
}

fn is_chorus(kind: &str) -> bool {
    kind.eq_ignore_ascii_case("chorus") || kind.eq_ignore_ascii_case("hook")
}

fn containing(cue: f64, segments: &[LyricSegment]) -> Option<&LyricSegment> {
    // A boundary timestamp belongs to the segment that starts there.
    segments
        .iter()
        .find(|segment| (segment.start - cue).abs() <= 1e-9)
        .or_else(|| {
            segments
                .iter()
                .find(|segment| cue + 1e-9 > segment.start && cue < segment.end - 1e-9)
        })
        .or_else(|| {
            segments
                .iter()
                .find(|segment| cue + 1e-9 >= segment.start && cue <= segment.end + 1e-9)
        })
}

/// Classify a cue against lyric segments. No segments → unknown.
#[must_use]
pub fn classify_cue(cue: f64, segments: &[LyricSegment]) -> CuePlacement {
    if segments.is_empty() {
        return CuePlacement::Unknown;
    }
    if cue <= START_TOLERANCE_SECONDS {
        let first = segments
            .iter()
            .min_by(|a, b| a.start.partial_cmp(&b.start).unwrap());
        if let Some(first) = first {
            if first.start >= INTRO_BEFORE_VERSE_SECONDS
                || cue + START_TOLERANCE_SECONDS < first.start
            {
                return CuePlacement::Intro;
            }
        } else {
            return CuePlacement::Intro;
        }
    }
    let Some(segment) = containing(cue, segments) else {
        let first_start = segments
            .iter()
            .map(|segment| segment.start)
            .fold(f64::INFINITY, f64::min);
        if cue + START_TOLERANCE_SECONDS < first_start {
            return CuePlacement::Intro;
        }
        return CuePlacement::Unknown;
    };
    let into = cue - segment.start;
    if is_verse(&segment.kind) {
        if into <= START_TOLERANCE_SECONDS {
            CuePlacement::VerseStart
        } else {
            CuePlacement::MidVerse
        }
    } else if is_chorus(&segment.kind) {
        if into <= START_TOLERANCE_SECONDS {
            CuePlacement::Chorus
        } else {
            CuePlacement::MidChorus
        }
    } else if cue <= START_TOLERANCE_SECONDS {
        CuePlacement::Intro
    } else {
        CuePlacement::Unknown
    }
}

fn verse_cut_by_window(start: f64, end: f64, segments: &[LyricSegment]) -> Option<&LyricSegment> {
    segments.iter().find(|item| {
        is_verse(&item.kind) && (item.end - item.start) <= MAX_VERSE_SECONDS && {
            let overlap_lo = start.max(item.start + START_TOLERANCE_SECONDS);
            let overlap_hi = end.min(item.end);
            overlap_hi > overlap_lo + 1e-9
        }
    })
}

/// False for instrumental-only / no-vocal beds — no artist verse exists.
#[must_use]
pub fn has_vocal_verses(title: &str, path: &str) -> bool {
    classify_stem(title, path) != StemKind::InstrumentalOnly
}

/// If `proposed_cue` is mid-verse, or the incoming blend *finishes* mid-verse,
/// move to 0:00 (iconic intro) or a pre-roll onto the verse start.
#[must_use]
pub fn respect_verse_entry(
    proposed_cue: f64,
    segments: &[LyricSegment],
    first_beat: f64,
    bpm: Option<f64>,
    blend_beats: u32,
    title: &str,
    path: &str,
) -> VerseCueDecision {
    let _ = first_beat;
    if (!title.is_empty() || !path.is_empty()) && !has_vocal_verses(title, path) {
        return VerseCueDecision {
            legal: true,
            placement: CuePlacement::Unknown,
            cue_seconds: proposed_cue.max(0.0),
            reason: "instrumental / no vocals — verse boundaries do not apply".to_string(),
            source: "unchanged".to_string(),
        };
    }
    let placement = classify_cue(proposed_cue, segments);
    let period = bpm.filter(|value| *value > 0.0).map(|value| 60.0 / value);
    let fader_at = period.map_or(proposed_cue, |p| {
        proposed_cue + f64::from(blend_beats.max(1)) * p
    });
    if proposed_cue <= START_TOLERANCE_SECONDS {
        return VerseCueDecision {
            legal: true,
            placement: if placement == CuePlacement::Unknown {
                CuePlacement::Intro
            } else {
                placement
            },
            cue_seconds: proposed_cue.max(0.0),
            reason: "starting from the top is a legal mix-in".to_string(),
            source: "unchanged".to_string(),
        };
    }
    let cut = verse_cut_by_window(proposed_cue, fader_at, segments);
    if cut.is_none() && placement != CuePlacement::MidVerse {
        return VerseCueDecision {
            legal: true,
            placement,
            cue_seconds: proposed_cue.max(0.0),
            reason: format!("{placement:?} is a legal mix-in"),
            source: "unchanged".to_string(),
        };
    }
    let Some(verse) = cut
        .or_else(|| containing(proposed_cue, segments).filter(|segment| is_verse(&segment.kind)))
    else {
        return VerseCueDecision {
            legal: true,
            placement,
            cue_seconds: proposed_cue.max(0.0),
            reason: "mid-verse but no containing verse segment".to_string(),
            source: "unchanged".to_string(),
        };
    };
    let has_intro = verse.start >= INTRO_BEFORE_VERSE_SECONDS
        || segments
            .iter()
            .any(|segment| segment.end <= verse.start + 1e-9 && segment.start < verse.start);
    if has_intro {
        return VerseCueDecision {
            legal: false,
            placement: CuePlacement::MidVerse,
            cue_seconds: 0.0,
            reason: format!(
                "incoming blend from {proposed_cue:.2}s finishes inside verse {:.2}–{:.2}; start from 0:00 so the blend covers the intro",
                verse.start, verse.end
            ),
            source: "intro_top".to_string(),
        };
    }
    let preroll = period.map_or(0.0, |p| f64::from(blend_beats.max(1)) * p);
    let cue = (verse.start - preroll).max(0.0);
    VerseCueDecision {
        legal: false,
        placement: CuePlacement::MidVerse,
        cue_seconds: cue,
        reason: format!(
            "incoming blend from {proposed_cue:.2}s finishes inside verse {:.2}–{:.2}; pre-roll onto verse start",
            verse.start, verse.end
        ),
        source: "verse_preroll".to_string(),
    }
}

/// Stretch an automatic ride so the whole outgoing blend sits after the verse.
/// Returns `(ride_beats, reason)`.
#[must_use]
pub fn respect_verse_exit(
    cue_seconds: f64,
    ride_beats: u32,
    bpm: f64,
    segments: &[LyricSegment],
    blend_beats: u32,
    title: &str,
    path: &str,
) -> (u32, String) {
    if (!title.is_empty() || !path.is_empty()) && !has_vocal_verses(title, path) {
        return (ride_beats, "unchanged".to_string());
    }
    if segments.is_empty() || bpm <= 0.0 {
        return (ride_beats, "unchanged".to_string());
    }
    let period = 60.0 / bpm;
    let fade_at = cue_seconds + f64::from(ride_beats) * period;
    let fade_done = fade_at + f64::from(blend_beats.max(1)) * period;
    let Some(verse) = verse_cut_by_window(fade_at, fade_done, segments) else {
        return (ride_beats, "unchanged".to_string());
    };
    let needed = ((verse.end - cue_seconds) / period).ceil().max(0.0) as u32;
    if needed <= ride_beats {
        return (ride_beats, "unchanged".to_string());
    }
    let capped = needed.min(ride_beats.saturating_add(128));
    (
        capped,
        format!(
            "blend {fade_at:.2}–{fade_done:.2}s eats verse {:.2}–{:.2}; extend ride {ride_beats} -> {capped} so the fade starts at the verse end",
            verse.start, verse.end
        ),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    fn verse(start: f64, end: f64) -> LyricSegment {
        LyricSegment {
            kind: "verse".to_string(),
            start,
            end,
        }
    }

    fn chorus(start: f64, end: f64) -> LyricSegment {
        LyricSegment {
            kind: "chorus".to_string(),
            start,
            end,
        }
    }

    #[test]
    fn on_fire_phrase_body_is_mid_verse_and_rewrites_to_zero() {
        let segments = vec![
            verse(3.60, 12.92),
            chorus(12.92, 27.75),
            verse(27.75, 73.30),
            chorus(73.30, 88.11),
        ];
        assert_eq!(classify_cue(41.35, &segments), CuePlacement::MidVerse);
        let decision = respect_verse_entry(41.35, &segments, 0.292, Some(95.0), 32, "", "");
        assert!(!decision.legal);
        assert_eq!(decision.cue_seconds, 0.0);
        assert_eq!(decision.source, "intro_top");
    }

    #[test]
    fn cue_at_zero_is_intro() {
        let segments = vec![chorus(12.92, 27.75), verse(32.88, 73.30)];
        assert_eq!(classify_cue(0.0, &segments), CuePlacement::Intro);
        let decision = respect_verse_entry(0.0, &segments, 0.0, Some(95.0), 32, "", "");
        assert!(decision.legal);
        assert_eq!(decision.source, "unchanged");
    }

    #[test]
    fn verse_start_within_tolerance_is_legal() {
        let segments = vec![verse(32.88, 73.30)];
        assert_eq!(classify_cue(32.88, &segments), CuePlacement::VerseStart);
        assert_eq!(classify_cue(34.0, &segments), CuePlacement::VerseStart);
        assert_eq!(classify_cue(36.0, &segments), CuePlacement::MidVerse);
    }

    #[test]
    fn no_intro_prerolls_onto_verse_start() {
        let segments = vec![verse(4.0, 40.0)];
        let decision = respect_verse_entry(20.0, &segments, 0.0, Some(90.0), 16, "", "");
        assert!(!decision.legal);
        assert_eq!(decision.source, "verse_preroll");
        // 16 beats at 90 BPM = 10.67s; 4.0 - 10.67 → 0
        assert_eq!(decision.cue_seconds, 0.0);
    }

    #[test]
    fn exit_extends_to_verse_end() {
        let segments = vec![verse(32.88, 73.30), chorus(73.30, 88.11)];
        // cue 0, 80 beats at 95 BPM = 50.53s, still in verse 1
        let (ride, reason) = respect_verse_exit(0.0, 80, 95.0, &segments, 32, "", "");
        assert!(ride > 80, "{ride} {reason}");
        assert!(reason.contains("extend ride"));
        let fade = f64::from(ride) * 60.0 / 95.0;
        assert!((fade - 73.30).abs() < 1.0, "fade {fade}");
    }

    #[test]
    fn twenty_one_questions_blend_in_rewrites_to_zero() {
        let segments = vec![
            verse(2.66, 32.77),
            chorus(32.77, 53.02),
            verse(53.02, 94.50),
        ];
        let decision = respect_verse_entry(52.76, &segments, 0.0, Some(93.05), 32, "", "");
        assert!(!decision.legal);
        assert_eq!(decision.cue_seconds, 0.0);
        assert_eq!(decision.source, "intro_top");
    }

    #[test]
    fn empty_segments_do_not_rewrite() {
        assert_eq!(classify_cue(41.35, &[]), CuePlacement::Unknown);
        let decision = respect_verse_entry(41.35, &[], 0.0, Some(95.0), 32, "", "");
        assert!(decision.legal);
        assert_eq!(decision.cue_seconds, 41.35);
        let (ride, reason) = respect_verse_exit(0.0, 80, 95.0, &[], 32, "", "");
        assert_eq!(ride, 80);
        assert_eq!(reason, "unchanged");
    }

    #[test]
    fn instrumental_skips_verse_rules_even_with_vocal_lrc() {
        let segments = vec![verse(27.75, 73.30), chorus(73.30, 88.11)];
        let decision = respect_verse_entry(
            41.35,
            &segments,
            0.0,
            Some(95.0),
            32,
            "On Fire (Instrumental)",
            "",
        );
        assert!(decision.legal);
        assert_eq!(decision.cue_seconds, 41.35);
        let (ride, reason) =
            respect_verse_exit(0.0, 80, 95.0, &segments, 32, "On Fire (Instrumental)", "");
        assert_eq!(ride, 80);
        assert_eq!(reason, "unchanged");
        assert!(!has_vocal_verses("On Fire (Instrumental)", ""));
        assert!(has_vocal_verses("On Fire (Feat. 50 Cent)", ""));
    }
}
